src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs

```diff
@@ -356,93 +447,427 @@ bool IsCreateHelper(IMethodSymbol method)
+                ReportDefineDiagnosticArgumentCorrectlyDiagnostic(DefineDiagnosticTitleCorrectlyRule,
+                    argumentOperation, fixedTitle, fixLocation, reportDiagnostic);
+            }
+        }
 
-            if (!TryGetNonEmptyConstantStringValue(titleArgument, out var title))
+        private static void ReportDefineDiagnosticArgumentCorrectlyDiagnostic(
+            DiagnosticDescriptor descriptor,
+            IArgumentOperation argumentOperation,
+            string fixValue,
+            Location fixLocation,
+            Action<Diagnostic> reportDiagnostic)
+        {
+            // Additional location in an additional document does not seem to be preserved
+            // from analyzer to code fix due to a Roslyn bug.
+            // We workaround this bug by passing additional document file path and location span as strings.
+            // TODO: File a Roslyn bug.
+
+            var additionalLocations = ImmutableArray<Location>.Empty;
+            var properties = ImmutableDictionary<string, string>.Empty.Add(DefineDescriptorArgumentCorrectlyFixValue, fixValue);
+            if (fixLocation.IsInSource)
             {
-                return;
+                additionalLocations = additionalLocations.Add(fixLocation);
             }
+            else
+            {
+                var span = fixLocation.SourceSpan;
+                var locationInfo = $"{span.Start};{span.Length};{fixLocation.GetLineSpan().Path}";
+                properties = properties.Add(DefineDescriptorArgumentCorrectlyFixAdditionalDocumentLocationInfo, locationInfo);
+            }
+
+            var location = argumentOperation.Syntax.GetLocation();
+            var diagnostic = Diagnostic.Create(descriptor, location, additionalLocations: additionalLocations, properties: properties);
+            reportDiagnostic(diagnostic);
+        }
+
+        internal static bool TryGetAdditionalDocumentLocationInfo(Diagnostic diagnostic,
+            [NotNullWhen(returnValue: true)] out string? filePath,
+            [NotNullWhen(returnValue: true)] out TextSpan? fileSpan)
+        {
+            Debug.Assert(diagnostic.Id == DiagnosticIds.DefineDiagnosticTitleCorrectlyRuleId ||
+                diagnostic.Id == DiagnosticIds.DefineDiagnosticMessageCorrectlyRuleId ||
+                diagnostic.Id == DiagnosticIds.DefineDiagnosticDescriptionCorrectlyRuleId);
 
-            if (IsMultiSentences(title) || EndsWithPeriod(title) || ContainsLineReturn(title))
+            if (diagnostic.Properties.TryGetValue(DefineDescriptorArgumentCorrectlyFixAdditionalDocumentLocationInfo, out var locationInfo))
             {
-                operationAnalysisContext.ReportDiagnostic(titleArgument.CreateDiagnostic(DefineDiagnosticTitleCorrectlyRule));
+                var index = locationInfo.IndexOf(';');
```

Comment:
[question] Maybe this is better for performance but wouldn't it be better to do a split and work with the strings rather than doing a locate, tryparse repeat approach?

Reply:
I am not sure if file names/paths on different OS can allow ';' character. Your approach also seems reasonable.

Reply:
Done.